NettyHandlerAdapter is noisy when portscraped over ssl

When running Ratpack (1.8.2) using HTTPS on a public facing linux server, any non-ssl requests (such as port scrapes, or misconfigured clients) cause one of the following exceptions to be logged at ERROR level.

```
r.s.i.NettyHandlerAdapter {}  io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: error:100000f0:SSL routines:OPENSSL_internal:UNSUPPORTED_PROTOCOL
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:471)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:795)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at ratpack.exec.internal.DefaultExecController$ExecControllerBindingThreadFactory.lambda$newThread$0(DefaultExecController.java:137)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: javax.net.ssl.SSLHandshakeException: error:100000f0:SSL routines:OPENSSL_internal:UNSUPPORTED_PROTOCOL
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.sslReadErrorResult(ReferenceCountedOpenSslEngine.java:1288)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1249)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1325)
	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1368)
	at io.netty.handler.ssl.SslHandler$SslEngineType$1.unwrap(SslHandler.java:206)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1387)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1282)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1329)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	... 16 more
```

Or

```
r.s.i.NettyHandlerAdapter {}  io.netty.handler.codec.DecoderException: io.netty.handler.ssl.NotSslRecordException: not an SSL/TLS record: 474554202f636f6e6669672f676574757365723f696e6465783d3020485454502f312e310d0a486f73743a20332e32352e3232322e3132343a3434330d0a557365722d4167656e743a204d6f7a696c6c612f352e3020285831313b205562756e74753b204c696e7578207838365f36343b2072763a37362e3029204765636b6f2f32303130303130312046697265666f782f37362e300d0a4163636570743a20746578742f68746d6c2c6170706c69636174696f6e2f7868746d6c2b786d6c2c6170706c69636174696f6e2f786d6c3b713d302e392c696d6167652f776562702c2a2f2a3b713d302e380d0a4163636570742d4c616e67756167653a20656e2d47422c656e3b713d302e350d0a4163636570742d456e636f64696e673a20677a69702c206465666c6174650d0a436f6e6e656374696f6e3a20636c6f73650d0a557067726164652d496e7365637572652d52657175657374733a20310d0a0d0a
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:471)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:795)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at ratpack.exec.internal.DefaultExecController$ExecControllerBindingThreadFactory.lambda$newThread$0(DefaultExecController.java:137)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.netty.handler.ssl.NotSslRecordException: not an SSL/TLS record: 474554202f636f6e6669672f676574757365723f696e6465783d3020485454502f312e310d0a486f73743a20332e32352e3232322e3132343a3434330d0a557365722d4167656e743a204d6f7a696c6c612f352e3020285831313b205562756e74753b204c696e7578207838365f36343b2072763a37362e3029204765636b6f2f32303130303130312046697265666f782f37362e300d0a4163636570743a20746578742f68746d6c2c6170706c69636174696f6e2f7868746d6c2b786d6c2c6170706c69636174696f6e2f786d6c3b713d302e392c696d6167652f776562702c2a2f2a3b713d302e380d0a4163636570742d4c616e67756167653a20656e2d47422c656e3b713d302e350d0a4163636570742d456e636f64696e673a20677a69702c206465666c6174650d0a436f6e6e656374696f6e3a20636c6f73650d0a557067726164652d496e7365637572652d52657175657374733a20310d0a0d0a
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1261)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1329)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	... 16 more
```

It would be nice to be able to customize Ratpack to suppress the stack trace, and modify this to be an `INFO` level message instead of an `ERROR`.  Obviously suppression is preferable to hiding this message, as sometimes it is useful for diagnosing when actual customers have a misconfiguration.  But the bots make it too noisy at present, and the level makes it trigger events in out log watching machinery

Cheers,

Tim

--- a/ratpack-core/ratpack-core.gradle
+++ b/ratpack-core/ratpack-core.gradle
@@ -63,6 +63,10 @@ dependencies {
 
   testImplementation commonDependencies.commonsLang3
 
+  // Used to do some testing of log output
+  testImplementation "org.apache.logging.log4j:log4j-api:${commonVersions.log4j}"
+  testImplementation "org.apache.logging.log4j:log4j-core:${commonVersions.log4j}"
+
   testImplementation('com.opencsv:opencsv:5.5.2') {
     exclude group: "org.apache.commons", module: "commons-lang3"
   }

--- a/ratpack-core/src/main/java/ratpack/core/config/internal/module/ServerConfigDataDeserializer.java
+++ b/ratpack-core/src/main/java/ratpack/core/config/internal/module/ServerConfigDataDeserializer.java
@@ -31,6 +31,7 @@ import io.netty.handler.ssl.SslContext;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import ratpack.config.FileSystemBinding;
+import ratpack.core.server.DecodingErrorLevel;
 import ratpack.core.server.ServerConfig;
 import ratpack.core.server.internal.ServerConfigData;
 import ratpack.core.server.internal.ServerEnvironment;
@@ -127,6 +128,9 @@ public class ServerConfigDataDeserializer extends JsonDeserializer<ServerConfigD
     }
     if (serverNode.hasNonNull("portFile")) {
       data.setPortFile(toValue(codec, serverNode.get("portFile"), Path.class));
+    }
+    if (serverNode.hasNonNull("decodingErrorLevel")) {
+      data.setDecodingErrorLevel(DecodingErrorLevel.valueOf(serverNode.get("decodingErrorLevel").asText("WARN")));
     }
 
     return data;

--- a/ratpack-core/src/main/java/ratpack/core/server/DecodingErrorLevel.java
+++ b/ratpack-core/src/main/java/ratpack/core/server/DecodingErrorLevel.java
@@ -1,0 +1,31 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ratpack.core.server;
+
+/**
+ * Specifies how messages related to request decoding are handled in the log output.
+ *
+ * @since 2.0
+ */
+public enum DecodingErrorLevel {
+
+  NONE,
+  FULL,
+  INFO,
+  WARN,
+  ERROR
+}

--- a/ratpack-core/src/main/java/ratpack/core/server/ServerConfig.java
+++ b/ratpack-core/src/main/java/ratpack/core/server/ServerConfig.java
@@ -375,4 +375,16 @@ public interface ServerConfig extends ConfigData {
    */
   FileSystemBinding getBaseDir() throws NoBaseDirException;
 
+  /**
+   * Specifies how to log messages associated with request decode failures.
+   * <p>
+   * This is useful when using Ratpack as an HTTPS server and misbehaving clients are sending HTTP requests.
+   * <p>
+   * The default value for this setting is {@code WARN} which emits a single line message for each non SSL/TLS record.
+   *
+   * @return the configured log setting
+   * @since 2.0
+   */
+  DecodingErrorLevel getDecodingErrorLevel();
+
 }

--- a/ratpack-core/src/main/java/ratpack/core/server/ServerConfigBuilder.java
+++ b/ratpack-core/src/main/java/ratpack/core/server/ServerConfigBuilder.java
@@ -598,4 +598,14 @@ public interface ServerConfigBuilder extends ConfigDataBuilder {
    */
   @Override
   ServerConfig build();
+
+  /**
+   * Configures how to emit messages associated with errors when decoding incoming requests.
+   *
+   * @param log the logging setting for failed decodings
+   * @return {@code this}
+   * @see ServerConfig#getDecodingErrorLevel() ()
+   * @since 2.0
+   */
+  ServerConfigBuilder decodingErrorLevel(DecodingErrorLevel log);
 }

--- a/ratpack-core/src/main/java/ratpack/core/server/internal/DefaultServerConfig.java
+++ b/ratpack-core/src/main/java/ratpack/core/server/internal/DefaultServerConfig.java
@@ -22,6 +22,7 @@ import ratpack.config.ConfigData;
 import ratpack.config.ConfigObject;
 import ratpack.config.internal.DelegatingConfigData;
 import ratpack.config.FileSystemBinding;
+import ratpack.core.server.DecodingErrorLevel;
 import ratpack.core.server.NoBaseDirException;
 import ratpack.core.server.ServerConfig;
 import ratpack.func.Nullable;
@@ -149,5 +150,10 @@ public class DefaultServerConfig extends DelegatingConfigData implements ServerC
   @Override
   public Optional<Path> getPortFile() {
     return serverConfigData.getPortFile();
+  }
+
+  @Override
+  public DecodingErrorLevel getDecodingErrorLevel() {
+    return serverConfigData.getDecodingErrorLevel();
   }
 }

--- a/ratpack-core/src/main/java/ratpack/core/server/internal/DefaultServerConfigBuilder.java
+++ b/ratpack-core/src/main/java/ratpack/core/server/internal/DefaultServerConfigBuilder.java
@@ -36,6 +36,7 @@ import ratpack.core.impose.ForceDevelopmentImposition;
 import ratpack.core.impose.ForceServerListenPortImposition;
 import ratpack.core.impose.Impositions;
 import ratpack.core.impose.ServerConfigImposition;
+import ratpack.core.server.DecodingErrorLevel;
 import ratpack.core.server.ServerConfig;
 import ratpack.core.server.ServerConfigBuilder;
 import ratpack.func.Action;
@@ -374,6 +375,11 @@ public class DefaultServerConfigBuilder implements ServerConfigBuilder {
   @Override
   public ImmutableList<ConfigSource> getConfigSources() {
     return configDataBuilder.getConfigSources();
+  }
+
+  @Override
+  public ServerConfigBuilder decodingErrorLevel(DecodingErrorLevel errorLog) {
+    return addToServer(n -> n.put("decodingErrorLevel", errorLog.name()));
   }
 
   @Override

--- a/ratpack-core/src/main/java/ratpack/core/server/internal/NettyHandlerAdapter.java
+++ b/ratpack-core/src/main/java/ratpack/core/server/internal/NettyHandlerAdapter.java
@@ -20,6 +20,7 @@ import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufUtil;
 import io.netty.buffer.Unpooled;
 import io.netty.channel.*;
+import io.netty.handler.codec.DecoderException;
 import io.netty.handler.codec.http.*;
 import io.netty.handler.ssl.SslHandler;
 import io.netty.handler.ssl.SslHandshakeCompletionEvent;
@@ -40,6 +41,7 @@ import ratpack.core.http.MutableHeaders;
 import ratpack.core.http.Response;
 import ratpack.core.http.internal.*;
 import ratpack.core.render.internal.DefaultRenderController;
+import ratpack.core.server.DecodingErrorLevel;
 import ratpack.core.server.ServerConfig;
 import ratpack.exec.ExecController;
 import ratpack.func.Action;
@@ -73,6 +75,7 @@ public class NettyHandlerAdapter extends ChannelInboundHandlerAdapter {
 
   private final Registry serverRegistry;
   private final boolean development;
+  private final DecodingErrorLevel decodingErrorLevel;
   private final Clock clock;
 
   public NettyHandlerAdapter(Registry serverRegistry, Handler handler) throws Exception {
@@ -80,6 +83,7 @@ public class NettyHandlerAdapter extends ChannelInboundHandlerAdapter {
     this.serverRegistry = serverRegistry;
     this.applicationConstants = new DefaultContext.ApplicationConstants(this.serverRegistry, new DefaultRenderController(), serverRegistry.get(ExecController.class), Handlers.notFound());
     this.development = serverRegistry.get(ServerConfig.class).isDevelopment();
+    this.decodingErrorLevel = serverRegistry.get(ServerConfig.class).getDecodingErrorLevel();
     this.clock = serverRegistry.get(Clock.class);
   }
 
@@ -234,7 +238,23 @@ public class NettyHandlerAdapter extends ChannelInboundHandlerAdapter {
   @Override
   public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
     if (!isIgnorableException(cause)) {
-      LOGGER.error("", cause);
+      if (cause instanceof DecoderException && cause.getCause() != null && decodingErrorLevel != DecodingErrorLevel.FULL) {
+        String msg = cause.getCause().getMessage();
+        switch (decodingErrorLevel) {
+          case INFO:
+            LOGGER.info(msg);
+            break;
+          case WARN:
+            LOGGER.warn(msg);
+            break;
+          case ERROR:
+            LOGGER.error(msg);
+            break;
+          default:
+        }
+      } else {
+        LOGGER.error("", cause);
+      }
       if (ctx.channel().isActive()) {
         sendError(ctx, HttpResponseStatus.INTERNAL_SERVER_ERROR);
       }

--- a/ratpack-core/src/main/java/ratpack/core/server/internal/ServerConfigData.java
+++ b/ratpack-core/src/main/java/ratpack/core/server/internal/ServerConfigData.java
@@ -18,6 +18,7 @@ package ratpack.core.server.internal;
 
 import io.netty.handler.ssl.SslContext;
 import ratpack.config.FileSystemBinding;
+import ratpack.core.server.DecodingErrorLevel;
 import ratpack.core.server.ServerConfig;
 
 import java.net.InetAddress;
@@ -48,6 +49,7 @@ public class ServerConfigData {
   private int maxChunkSize = ServerConfig.DEFAULT_MAX_CHUNK_SIZE;
   private int maxInitialLineLength = ServerConfig.DEFAULT_MAX_INITIAL_LINE_LENGTH;
   private int maxHeaderSize = ServerConfig.DEFAULT_MAX_HEADER_SIZE;
+  private DecodingErrorLevel decodingErrorLevel = DecodingErrorLevel.WARN;
 
   private Duration idleTimeout = Duration.ZERO;
   private boolean registerShutdownHook = true;
@@ -228,5 +230,13 @@ public class ServerConfigData {
 
   public void setRegisterShutdownHook(boolean registerShutdownHook) {
     this.registerShutdownHook = registerShutdownHook;
+  }
+
+  public DecodingErrorLevel getDecodingErrorLevel() {
+    return decodingErrorLevel;
+  }
+
+  public void setDecodingErrorLevel(DecodingErrorLevel decoderErrorLog) {
+    this.decodingErrorLevel = decoderErrorLog;
   }
 }

--- a/ratpack-core/src/test/groovy/ratpack/core/ssl/HttpsSpec.groovy
+++ b/ratpack-core/src/test/groovy/ratpack/core/ssl/HttpsSpec.groovy
@@ -17,10 +17,15 @@
 package ratpack.core.ssl
 
 import com.google.common.base.Throwables
+import io.netty.handler.codec.PrematureChannelClosureException
 import io.netty.handler.ssl.ClientAuth
 import io.netty.handler.ssl.SslContext
 import io.netty.handler.ssl.SslContextBuilder
 import io.netty.handler.ssl.util.SelfSignedCertificate
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.core.Logger
+import org.apache.logging.log4j.core.appender.WriterAppender
+import ratpack.core.server.internal.NettyHandlerAdapter
 import ratpack.test.internal.RatpackGroovyDslSpec
 import spock.lang.Shared
 import sun.security.provider.certpath.SunCertPathBuilderException
@@ -134,5 +139,51 @@ class HttpsSpec extends RatpackGroovyDslSpec {
 
     then:
     requestSpec { it.sslContext(clientContext) }.text == nonLocalHostCert.cert().subjectDN.name
+  }
+
+  def "suppresses exceptions from non-HTTPS request"() {
+    given:
+    // Look up and modifying the underlying log4j configuration
+    // so we can capture the log output in this test
+    Logger logs = (Logger) LogManager.getLogger(NettyHandlerAdapter)
+    def sw = new StringWriter()
+
+    def builder = WriterAppender.newBuilder()
+    builder.setTarget(sw)
+    builder.setName("testCapture")
+    def testAppender = builder.build()
+    testAppender.start()
+
+    logs.setAdditive(true)
+    logs.addAppender(testAppender)
+
+    SslContext localhostContext = SslContextBuilder.forServer(localHostCert.certificate(), localHostCert.privateKey()).build()
+    serverConfig {
+      ssl(localhostContext)
+    }
+
+    when:
+    handlers {
+      get {
+        render "ok"
+      }
+    }
+
+    then:
+    try {
+      getText("http://localhost:${applicationUnderTest.address.port}")
+    } catch(PrematureChannelClosureException e) {
+      assert e.getMessage().contains("closed the connection prematurely")
+    }
+
+    and:
+    def lines = sw.toString().trim().split("\n")
+    lines.size() == 1
+    lines[0].startsWith("not an SSL/TLS record")
+
+    cleanup:
+    testAppender.stop()
+    logs.removeAppender(testAppender)
+    logs.setAdditive(false)
   }
 }

